### PR TITLE
Add OpenTelemetry attributes to the metrics backend

### LINKED ISSUE #2292: Add OpenTelemetry attributes to the metrics backend
Currently the [OpenTelemetryMetricsBackend](https://github.com/softwaremill/sttp/blob/master/observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala) reports a number of metrics, following the naming conventions of OpenTelemetry. 

However, we only report raw metric values, without any of the attributes as defined [here](https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#http-client).

The goal of this issue is to report the metrics along with the **required** attributes. Optionally, we might include other attributes, if this is feasible and "makes sense".

The abstractions currently present in `OpenTelemetryMetricsBackend` might need updating/extending/refactoring; since this targets version 4 (yet unreleased), there are no compatibility requirements.

### PR DESCRIPTION
**Pull request by varshith257**
Before submitting pull request:
- [x] Check if the project compiles by running `sbt compile`
- [x] Verify docs compilation by running `sbt compileDocs`
- [x] Check if tests pass by running `sbt test`
- [x] Format code by running `sbt scalafmt`

Closes #2292 
/claim #2292

### CODE REVIEW

**Comment by varshith257** [comment]
@adamw Have a review on it at your free time

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala [hunk 1]
```diff
@@ -195,6 +211,32 @@ private class OpenTelemetryMetricsListener(
     b = config.description.fold(b)(b.setDescription)
     b.build()
   }
+
+  private def createRequestAttributes(request: GenericRequest[_, _]): Attributes = {
+    val attributes = Attributes
```
Comment (adamw) [inline observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:216]: maybe you could add links to appropriate section of the OT specification, just for reference why we're setting these specific attributes?

Comment (adamw) [reply observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:216]: (as comments)

Comment (adamw) [reply observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:216]: or simply a single link to the page listing the attributes

File: observability/opentelemetry-metrics-backend/src/test/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackendTest.scala [hunk 2]
```diff
@@ -177,7 +178,11 @@ class OpenTelemetryMetricsBackendTest extends AnyFlatSpec with Matchers with Opt
     (0 until 5).foreach(_ => basicRequest.get(uri"http://127.0.0.1/foo").send(backend))
 
     // then
-    getHistogramValue(reader, OpenTelemetryMetricsBackend.DefaultLatencyHistogramName).map(_.getSum) should not be empty
+    val histogram = getHistogramValue(reader, OpenTelemetryMetricsBackend.DefaultLatencyHistogramName).value
+    val attributes = histogram.getAttributes
+    attributes.get(AttributeKey.stringKey("http.request.method")) shouldBe "GET"
```
Comment (adamw) [inline observability/opentelemetry-metrics-backend/src/test/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackendTest.scala:183]: can we maybe use `specTest` here as well?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,7 +1,6 @@
 name: CI
 on:
   pull_request:
-    branches: ['**']
```
Comment (adamw) [inline .github/workflows/ci.yml:4]: was this removed by mistake? I think it should stay as-is?

**Comment by adamw** [comment]
Thanks - the PR looks good! I've left a couple of minor comments before we merge

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala [hunk 4]
```diff
@@ -195,6 +211,32 @@ private class OpenTelemetryMetricsListener(
...
+      .builder()
+      .put(AttributeKey.stringKey("http.request.method"), request.method.method)
+      .put(AttributeKey.stringKey("server.address"), request.uri.host.getOrElse("unknown"))
+      .put(AttributeKey.longKey("server.port").asInstanceOf[AttributeKey[Any]], request.uri.port.getOrElse(80).toLong)
```
Comment (sergiuszkierat) [inline observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:220]: can we use a simpler version without casting ? :thinking: 
`.put(AttributeKey.longKey("server.port"), request.uri.port.getOrElse(80))`

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala [hunk 5]
```diff
@@ -195,6 +211,32 @@ private class OpenTelemetryMetricsListener(
...
+      .build()
+
+    attributes
+  }
+
+  private def createResponseAttributes(response: Response[_]): Attributes =
+    Attributes
+      .builder()
+      .put(AttributeKey.longKey("http.response.status_code").asInstanceOf[AttributeKey[Any]], response.code.code.toLong)
```
Comment (sergiuszkierat) [inline observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:229]: what about `.put(AttributeKey.longKey("http.response.status_code"), response.code.code)` ? :thinking:

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala [hunk 6]
```diff
@@ -108,44 +108,60 @@ private class OpenTelemetryMetricsListener(
...
+  private def recordHistogram(
+      config: Option[HistogramCollectorConfig],
+      size: Option[Long],
+      attributes: Attributes
+  ): Unit = config.foreach { cfg =>
+    getOrCreateHistogram(histograms, cfg, createNewHistogram).record(size.getOrElse(0L).toDouble, attributes)
   }
 
-  private def incrementCounter(collectorConfig: Option[CollectorConfig]): Unit =
+  private def incrementCounter(collectorConfig: Option[CollectorConfig], attributes: Attributes): Unit =
```
Comment (sergiuszkierat) [inline observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:164]: `attributes: Attributes` never used :baby_chick:

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala (hunk 4 above)
Comment (varshith257) [reply observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:220]: Yeah! I just added typecast initially to look it's behavior and left it.

File: observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala (hunk 5 above)
Comment (varshith257) [reply observability/opentelemetry-metrics-backend/src/main/scala/sttp/client4/opentelemetry/OpenTelemetryMetricsBackend.scala:229]: Thanks! Done

**Comment by varshith257** [comment]
@adamw Done! Added review comments

**Comment by varshith257** [comment]
@adamw Can it be merged now?

**Comment by adamw** [comment]
@varshith257 Yes, all done, thanks :)

**Comment by varshith257** [comment]
@adamw Any communication channel to connect with you? I have some doubts regarding implementation of other issue works

**Comment by adamw** [comment]
@varshith257 I think it would be best to keep the discussions on the appropriate GH issue, otherwise there's softwaremill.community Discourse and Scala Space's Discord

**Pull request merged**